core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,327 @@
+
+        @Nullable
+        private final Throwable cause;
+
+        /**
+         * No need to be volatile because updated only by the {@link #executor()}.
+         */
+        private int numAttemptsSoFar = 1;
+
+        private volatile boolean servedFromCache;
+
+        @VisibleForTesting
+        @Nullable
+        ScheduledFuture<?> refreshFuture;
+
+        CacheEntry(@Nullable InetAddress address, long ttlMillis, List<DnsQuestion> questions,
+                   @Nullable Throwable cause) {
+            this.address = address;
+            this.ttlMillis = ttlMillis;
+            this.questions = questions;
+            this.cause = cause;
+        }
+
+        void servedFromCache() {
+            servedFromCache = true;
+        }
+
+        @Nullable
+        InetAddress address() {
+            return address;
+        }
+
+        long ttlMillis() {
+            return ttlMillis;
+        }
+
+        @Nullable
+        Throwable cause() {
+            return cause;
+        }
+
+        void scheduleRefresh(long nextDelayMillis) {
+            if (resolverClosed) {
+                return;
+            }
+            refreshFuture = executor().schedule(this, nextDelayMillis, TimeUnit.MILLISECONDS);
+        }
+
+        void clear() {
+            assert resolverClosed;
+            if (refreshFuture != null) {
+                refreshFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (resolverClosed) {
+                return;
+            }
+
+            assert address != null;
+            final String hostName = address.getHostName();
+            if (!servedFromCache) {
+                cache.remove(hostName);
+                return;
+            }
+
+            final CompletableFuture<CacheEntry> result = new CompletableFuture<>();
+            sendQueries(questions, hostName, result);
+            result.handle((entry, cause) -> {
+                if (resolverClosed) {
+                    return null;
+                }
+
+                if (cause != null) {
+                    final long nextDelayMillis = refreshBackoff.nextDelayMillis(numAttemptsSoFar++);
+                    if (nextDelayMillis < 0) {
+                        cache.remove(hostName);
+                        return null;
+                    }
+                    scheduleRefresh(nextDelayMillis);
+                    return null;
+                }
+
+                // Got the response successfully so reset the state.
+                servedFromCache = true;
+                numAttemptsSoFar = 1;
+
+                if (entry.address().equals(address) && entry.ttlMillis() == ttlMillis) {
+                    scheduleRefresh(ttlMillis);
+                } else {
+                    // Replace the old entry with the new one.
+                    cache.put(hostName, result);
+                    entry.scheduleRefresh(entry.ttlMillis());
+                }
```

Comment:
I have a question on `entry.scheduleRefresh(ttlMillis)` logic. The refreshing `entry` is scheduled in the end of `sendQueries(...)`. 
And there are different refreshing logics here. IIUC, doesn't it cause double scheduling?

Reply:
Ah yes, thanks for finding this. 😄